esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py

```diff
@@ -0,0 +1,374 @@
+        the weights_matrix contains the respective model weights in the
+        model_ensemble dimension based on this perfect model.
+
+        Special feature: nan values in the model_ensemble dimension will lead
+        to the model being excluded from the weights calculation. This is
+        is always the case for the perfect model itself (diagonal of the
+        matrix) but might also be the case for other models. This is
+        particularly important for a correct calculation of the independence
+        weighting.
+    percentiles : array_like, shape (2,)
+        Lower and upper percentile to use in the confidence test. Has to
+        satisfy 0 <= percentiles <=1 and percentiles[0] < percentiles[1]
+    weighted : bool, optional
+        If weighted is set to False (default: True) all values in the weights
+        matrix will be set to 1, except for nan values which will be preserved.
+        This can be used to calculate the unweighted baseline for each perfect
+        model, consistent with the weighted case.
+
+    Returns
+    -------
+    percentile_spread : array_like, shape (N,)
+        Full range spanned by the two percentiles for each perfect model.
+    inside_ratio: float
+        Ratio of perfect models inside their respective percentile_spread.
+    """
+    percentiles = xr.DataArray(percentiles, dims='percentile')
+
+    # need to rename for the inside_count comparison
+    target_perfect = target.rename(
+        {'model_ensemble': 'perfect_model_ensemble'})
+
+    if not weighted:  # replace with equal weight but keep nans
+        weights_matrix = 0 * weights_matrix + 1
+
+    percentiles_data = xr.apply_ufunc(
+        weighted_quantile,
+        target,
+        percentiles,
+        weights_matrix,
+        input_core_dims=[['model_ensemble'], ['percentile'],
+                         ['model_ensemble']],
+        output_core_dims=[['percentile']],
+        vectorize=True,
+    )
+
+    inside_count = xr.ufuncs.logical_and(
+        target_perfect >= percentiles_data.isel(percentile=0),
+        target_perfect <= percentiles_data.isel(percentile=1)).values
+    inside_ratio = inside_count.sum() / len(inside_count)
+
+    percentiles_spread = (percentiles_data.isel(percentile=1) -
+                          percentiles_data.isel(percentile=0))
+
+    return percentiles_spread, inside_ratio
+
+
+def optimize_confidence(target: 'xr.DataArray', weights_matrix: 'xr.DataArray',
+                        performance_sigma: float) -> float:
+    """Optimize the performance sigma for confidence.
+
+    Parameters
+    ----------
+    target : array_like, shape (N,)
+        See calculate_percentiles for more information.
+    weights_matrix : array_like, shape (N, N)
+        See calculate_percentiles for more information.
+    performance_sigma : float
+        The performance sigma value used to calculate the weights.
+
+    Returns
+    -------
+    cost_function_value : float
```

Comment:
I think it would be super useful if we could write a symbolic representation of the actual cost function here. Something like

```md
This function returns the value of the following cost function:

.. math::

    \begin{cases}
        \frac{x^2-x}{x},& \text{if } x\geq 1\\
        0,              & \text{otherwise}
    \end{cases}

Reply:
I redesigned  the cost function a bit after some discussion with @ruthlorenz . Testing it now.